stubs/setuptools/setuptools/__init__.pyi

```diff
@@ -45,6 +46,14 @@ __all__ = [
 
 __version__: str
 
+@type_check_only
+class _BuildInfo(TypedDict):
+    sources: list[str] | tuple[str]
+    obj_deps: NotRequired[dict[str, list[str] | tuple[str]]]
+    macros: NotRequired[list[tuple[str, str] | tuple[str]]]
+    include_dirs: NotRequired[list[str]]
+    cflags: NotRequired[list[str]]
```

Comment:
Normally I'd prefer defining such types close to where their implementation stands (so with the appropriate `Command` or `Distribution`) but in this case it'd end up in `distutils`/`setuptools/_distutils` so I guess in `setuptools.__init__` is fine.

I do have a concern using a type-only TypedDict with required members as a parameter: It makes it annoying to use a non-literal dict. Maybe that's ok in setuptools due to the nature of it looking like a config file, it shouldn't be much more complex than my example below?

It's possible, but requires a private type-only import, which isn't great:
```py
from typing import cast, TYPE_CHECKING
if TYPE_CHECKING:
    from setuptools import _BuildInfo

libs = [
    ("lib1", {"sources": []}),
    ("lib2", {"sources": []}),
]
setup(libraries=libs)  # Type error in both mypy and pyright due to missing "sources"
# Ok, but must quote type to avoid runtime error
setup(libraries=cast(list[tuple[str, _BuildInfo]], libs))
# Ok
setup(libraries=[("literal", {"sources": []})])
# Ok
libs_annotated: list[tuple[str, _BuildInfo]] = [ # Must quote type or use from __future__ import annotations
    ("lib1", {"sources": []}),
    ("lib2", {"sources": []}),
]
setup(libraries=libs_annotated)
```

Reply:
Yeah, perhaps it's not ideal but I'm not seeing a great alternative, obviously passing things in-line solves this problem, or just:

```python
libs: "list[tuple[str, _BuildInfo]]" = [
    ("lib1", {"sources": []}),
    ("lib2", {"sources": []}),
]
setup(libraries=libs)
```

Which would work without quotes given deferred annotations.

Reply:
At the same time, the kind of people to type-check their setup.py are probably the same going all out on typing (looking at myself here ^^)